### PR TITLE
Add Fast Digitization Algorithm

### PR DESCRIPTION
**Pull request by vvolkl**
Addresses #198 .

Adds an algorithm that smears the `TrackHits` according to the segmentation cell size and outputs `PositionedHits`.

### CODE REVIEW

File: Reconstruction/RecTracker/src/components/FastGaussSmearDigi.cpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+};
+
+/// fill vector with hits ordered according to the CellIdOrder struct
+void sortTrackHits(const fcc::TrackHitCollection* unsortedHits, std::vector<fcc::TrackHit>& sortedHits,
+                   DD4hep::DDSegmentation::BitField64* decoder) {
+  sortedHits.reserve(unsortedHits->size());
+  for (const auto& hit : *unsortedHits) {
+    sortedHits.push_back(hit);
+  }
+  std::sort(sortedHits.begin(), sortedHits.end(), CellIdOrder(decoder));
```
Comment (jlingema) [inline Reconstruction/RecTracker/src/components/FastGaussSmearDigi.cpp]: I guess the modern / more compact thing to do here would be to use a lambda, but I think this is fine as well.

File: Reconstruction/RecTracker/src/components/FastGaussSmearDigi.cpp [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+FastGaussSmearDigi::FastGaussSmearDigi(const std::string& name, ISvcLocator* svcLoc) : GaudiAlgorithm(name, svcLoc) {
+
+  declareProperty("smearedHits", m_smearedTrackHits, "Smeared Tracker hits (Output)");
+  declareProperty("trackHits", m_trackHits, "Tracker hits (Input)");
+  declareProperty("readoutName", m_readoutName, "Readout name");
```
Comment (jlingema) [inline Reconstruction/RecTracker/src/components/FastGaussSmearDigi.cpp]: You already declare this as the property `readoutNames` in the header.

File: Reconstruction/RecTracker/src/components/FastGaussSmearDigi.cpp [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+  StatusCode sc = GaudiAlgorithm::initialize();
+  if (sc.isFailure()) return sc;
+
+  auto lcdd = m_geoSvc->lcdd();
+  auto readout = lcdd->readout(m_readoutName);
+  m_decoder = readout.idSpec().decoder();
+  auto segmentationXZ = dynamic_cast<DD4hep::DDSegmentation::CartesianGridXZ*>(readout.segmentation().segmentation());
+  m_segGridSizeX = segmentationXZ->gridSizeX();
+  m_segGridSizeZ = segmentationXZ->gridSizeZ();
```
Comment (jlingema) [inline Reconstruction/RecTracker/src/components/FastGaussSmearDigi.cpp]: I guess you should handle here that the `dynamic_cast` actually worked by checking that `segmentationXZ` isn't a `nullptr`

File: Reconstruction/RecTracker/src/components/FastGaussSmearDigi.cpp [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+  auto edmPositions = m_smearedTrackHits.createAndPut();
+  for (const auto& hit : sortedHits) {
+    const fcc::BareHit& hitCore = hit.core();
+    m_decoder->setValue(hit.core().cellId);
+    /// the local coordinates on the module
+    // add segmentation info and smearing here
+    double l[3] = {(*m_decoder)["x"] * m_segGridSizeX + m_gaussDist() / sqrt(12.) * m_segGridSizeX, 0,
+                   (*m_decoder)["z"] * m_segGridSizeZ + m_gaussDist() / sqrt(12.) * m_segGridSizeZ};
+    // global coordinates, will be filled by the transform
+    double g[3] = {0, 0, 0};
```
Comment (jlingema) [inline Reconstruction/RecTracker/src/components/FastGaussSmearDigi.cpp]: Please don't use c-style arrays instead use `std::array`...

File: Reconstruction/RecTracker/src/components/FastGaussSmearDigi.cpp [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    // direct lookup of transformation in the volume manager is broken in dd4hep
+    auto detelement = m_volman.lookupDetElement(m_decoder->getValue());
+    const auto& localToGlobal = detelement.worldTransformation();
+    localToGlobal.LocalToMaster(l, g);
+    auto position = fcc::Point();
+    // default dd4hep units differ from fcc ones
+    position.x = g[0] * CM_2_MM;
+    position.y = g[1] * CM_2_MM;
+    position.z = g[2] * CM_2_MM;
+    edmPositions->create(position, hitCore);
```
Comment (jlingema) [inline Reconstruction/RecTracker/src/components/FastGaussSmearDigi.cpp]: ... and `std::array::data` here.

File: Reconstruction/RecTracker/src/components/FastGaussSmearDigi.cpp [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+  auto edmPositions = m_smearedTrackHits.createAndPut();
+  for (const auto& hit : sortedHits) {
+    const fcc::BareHit& hitCore = hit.core();
+    m_decoder->setValue(hit.core().cellId);
+    /// the local coordinates on the module
+    // add segmentation info and smearing here
+    std::array<double, 3> l = {(*m_decoder)["x"] * m_segGridSizeX + m_gaussDist() / sqrt(12.) * m_segGridSizeX, 0,
+                   (*m_decoder)["z"] * m_segGridSizeZ + m_gaussDist() / sqrt(12.) * m_segGridSizeZ};
+    // global coordinates, will be filled by the transform
+    std::array<double, 3> g = {0, 0, 0};
```
Comment (jlingema) [inline Reconstruction/RecTracker/src/components/FastGaussSmearDigi.cpp]: One last thing: Can we rename those to have something more meaningful? E.g. `l` -> `localPos` `g` -> `globalPos`?

Comment (vvolkl) [inline Reconstruction/RecTracker/src/components/FastGaussSmearDigi.cpp]: Done. Thanks for the comments!

**Pull request merged**